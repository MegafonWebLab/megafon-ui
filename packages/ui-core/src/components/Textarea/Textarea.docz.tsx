import * as React from 'react';

export const TextAreaWrapper: React.FC = ({ children }) => (
    <div style={{ width: 250 }}>{children}</div>
);
