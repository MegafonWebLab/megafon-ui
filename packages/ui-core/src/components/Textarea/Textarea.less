@import '~styles/base.less';

@b: e('.mfui-textarea');

@{b} {
    color: @soggyAsphalt;
    font-size: 15px;
    line-height: 20px;

    &__field {
        font-family: inherit;
        position: relative;
        z-index: 1;
        font-size: 15px;
        resize: none;
        height: 152px;
        width: 100%;
        box-sizing: border-box;
        outline: none;
        cursor: pointer;
        padding: 13px 20px 13px 20px;
        line-height: 20px;
        background-color: transparent;
        opacity: 1;
        border-radius: 0;
        -webkit-appearance: none;
        box-shadow: none;
    }
    &__field-wrapper {
        position: relative;
        flex: 1 0 100px;
        background: @clearWhite;
    }

    &__text {
        font-size: 12px;
        line-height: 16px;
        margin-top: 4px;
    }
}

//mod color:
@{b} {
    &_color_default & {
        &__field {
            color: @soggyAsphalt;
            border: 1px solid @spbSky2;
            background-color: transparent;
        }
        .no-touchevents &__field-wrapper:hover @{b}__field,
        &__field:focus {
            border-color: @soggyAsphalt;
        }
    }
    &_color_default&_valid & {
        .no-touchevents &__field-wrapper:hover @{b}__field,
        &__field {
            border-color: @green;
        }
    }
    &_color_default&_error & {
        &__field-wrapper:hover @{b}__field,
        &__field {
            border-color: @fury;
        }
        &__text_error {
            color: @fury;
        }
    }

    &_color_white & {
        &__field {
            color: @soggyAsphalt;
            border: 1px solid @clearWhite;
        }
        &__field-wrapper:hover @{b}__field,
        &__field:focus {
            border-color: @soggyAsphalt;
        }
    }
    &_color_white&_error & {
        &__field-wrapper:hover @{b}__field,
        &__field {
            border-color: @fury;
        }
        &__text_error {
            color: @fury;
        }
    }
    &__text_success {
        color: @green;
    }

}

//mod disabled
@{b} {
    &_disabled & {
        &__field-wrapper {
            background-color: @spbSky1 !important;
        }
        &__field {
            border-color: @spbSky1 !important;
            cursor: default !important;
            color: @soggyAsphalt !important;
            -webkit-text-fill-color: @soggyAsphalt !important;
        }
    }
}
