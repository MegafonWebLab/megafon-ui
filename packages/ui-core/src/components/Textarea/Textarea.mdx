---
name: Textarea
menu: Components
---
import { Playground, PropsTable } from 'docz';
import Textarea from './Textarea';

export const wrapper = (width = 250) => ({ children }) => {
    return (
        <div style={{width: `${width}px`}}>
            {children}
        </div>
    )
}

# Textarea

## Properties

<PropsTable of={Textarea} />

## Basic usage

<Playground wrapper={wrapper()}>
    <Textarea name="name"/>
</Playground>

## Error usage

<Playground wrapper={wrapper()}>
    <Textarea error={true} noticeText="Name is required" />
</Playground>

## Valid usage

<Playground wrapper={wrapper()}>
    <Textarea valid={true} />
</Playground>

## Comment usage

<Playground wrapper={wrapper()}>
    <Textarea commentText="Comment" />
</Playground>

## Success text usage

<Playground wrapper={wrapper()}>
    <Textarea successText="Field passed validation!" />
</Playground>

## Disable usage

<Playground wrapper={wrapper()}>
    <Textarea disabled={true} />
</Playground>
