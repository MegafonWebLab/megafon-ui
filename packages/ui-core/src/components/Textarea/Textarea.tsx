import * as React from 'react';
import * as PropTypes from 'prop-types';
import './Textarea.less';
import cnCreate from 'utils/cn';
import * as equal from 'deep-equal';
import detectTouch from 'utils/detectTouch';
import InputLabel from 'components/InputLabel/InputLabel';

interface ITextareaProps {
    /** Field title */
    label?: React.ReactNode;
    /** Field color scheme */
    color?: 'default' | 'white';
    /** Error/Notice text */
    noticeText?: string;
    /** Comment text */
    commentText?: string;
    /** Success text */
    successText?: string;
    /** Autofocus */
    autoFocus?: boolean;
    /** Autocomplete */
    autocomplete?: 'on' | 'off';
    /** Validation passed */
    valid?: boolean;
    /** Validation error */
    error?: boolean;
    /** Disable field */
    disabled?: boolean;
    /** Required field */
    required?: boolean;
    /** Field name */
    name?: string;
    /** Placeholder */
    placeholder?: string;
    /** Html id attribute */
    id?: string;
    /** Field value */
    value?: string;
    /** Max length  */
    maxLength?: number;
    /** Default value */
    defaultValue?: string;
    /** Custom classname */
    className?: string;
    /** Change handler */
    onChange?(e: React.SyntheticEvent<EventTarget>): void;
    /** Blur handler */
    onBlur?(e: React.SyntheticEvent<EventTarget>): void;
    /** Focus handler */
    onFocus?(e: React.SyntheticEvent<EventTarget>): void;
    /** KeyUp handler */
    onKeyUp?(e: React.SyntheticEvent<EventTarget>): void;
}

const cn = cnCreate('mfui-textarea');
class Textarea extends React.Component<ITextareaProps, {}> {
    static propTypes = {
        label: PropTypes.node,
        noticeText: PropTypes.string,
        commentText: PropTypes.string,
        successText: PropTypes.string,
        autoFocus: PropTypes.bool,
        autocomplete: PropTypes.oneOf(['on', 'off']),
        valid: PropTypes.bool,
        error: PropTypes.bool,
        disabled: PropTypes.bool,
        required: PropTypes.bool,
        name: PropTypes.string,
        placeholder: PropTypes.string,
        id: PropTypes.string,
        value: PropTypes.string,
        maxLength: PropTypes.number,
        defaultValue: PropTypes.string,
        className: PropTypes.string,
        onChange: PropTypes.func,
        onBlur: PropTypes.func,
        onFocus: PropTypes.func,
        onKeyUp: PropTypes.func,
    };

    static defaultProps: Partial<ITextareaProps> = {
        autocomplete: 'off',
        color: 'default',
    };

    inputNode: any = React.createRef();
    isTouch: boolean = detectTouch();

    shouldComponentUpdate(nextProps: ITextareaProps) {
        return !equal(this.props, nextProps);
    }

    blur = () => {
        this.inputNode.current.blur();
    }

    focus = () => {
        this.inputNode.current.focus();
    }

    renderElem() {
        const params = {
            disabled: this.props.disabled,
            name: this.props.name,
            id: this.props.id,
            placeholder: this.props.placeholder,
            onChange: this.props.onChange,
            onBlur: this.props.onBlur,
            onFocus: this.props.onFocus,
            onKeyUp: this.props.onKeyUp,
            autoFocus: this.props.autoFocus,
            defaultValue: this.props.defaultValue,
            value: this.props.value,
            required: this.props.required,
            autoComplete: this.props.autocomplete,
            className: cn('field'),
        };

        return <textarea ref={this.inputNode} {...params} />;
    }

    render() {
        const {
            error, color, valid, disabled,
            className, commentText, successText,
            noticeText, label, id,
        } = this.props;

        return (
            <div
                className={cn('', {
                    valid: valid,
                    error: error,
                    disabled,
                    color,
                }, className)}
            >
                {label && <InputLabel htmlFor={id}>{label}</InputLabel>}
                <div className={cn('field-wrapper', { 'no-touch': !this.isTouch })}>
                    <div>{this.renderElem()}</div>
                </div>
                {(error || valid) && noticeText &&
                    <div className={cn('text', { error: true })}>{noticeText}</div>}
                {commentText &&
                    <div className={cn('text', { comment: true })}>{commentText}</div>}
                {successText &&
                    <div className={cn('text', { success: true })}>{successText}</div>}
            </div>
        );
    }
}

export default Textarea;
