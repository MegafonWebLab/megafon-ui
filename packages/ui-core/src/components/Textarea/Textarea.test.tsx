import * as React from 'react';
import { shallow, mount } from 'enzyme';
import Textarea from './Textarea';

describe('<Textarea />', () => {
    it('it renders Textarea', () => {
        const wrapper = shallow(<Textarea />);
        expect(wrapper).toMatchSnapshot();
    });

    it('it renders texts', () => {
        const wrapper = shallow(<Textarea commentText="text" successText="text" />);
        expect(wrapper).toMatchSnapshot();
    });

    it('it renders without valid icon', () => {
        const wrapper = shallow(<Textarea valid={true} />);
        expect(wrapper).toMatchSnapshot();
    });

    it('it renders without error icon', () => {
        const wrapper = shallow(<Textarea error={true} />);
        expect(wrapper).toMatchSnapshot();
    });

    it('it renders error icon', () => {
        const wrapper = shallow(<Textarea error={true} noticeText="text" />);
        expect(wrapper).toMatchSnapshot();
    });

    it('it renders valid icon', () => {
        const wrapper = shallow(<Textarea valid={true} noticeText="text" />);
        expect(wrapper).toMatchSnapshot();
    });

    it('is checked shouldComponentUpdate', () => {
        const wrapper = shallow(<Textarea error={true} noticeText="text" />);

        wrapper.setProps({ error: true, noticeText: 'text' });
        expect(wrapper).toMatchSnapshot();
    });

    it('is checked shouldComponentUpdate with new props', () => {
        const wrapper = shallow(<Textarea error={true} noticeText="text" />);

        wrapper.setProps({ error: false, noticeText: 'text' });
        expect(wrapper).toMatchSnapshot();
    });

    it('is checked focus method', () => {
        const wrapper = mount(<Textarea />);
        const instance = wrapper.instance() as Textarea;

        expect(() => instance.focus()).not.toThrow();
    });

    it('is checked blur method', () => {
        const wrapper = mount(<Textarea />);
        const instance = wrapper.instance() as Textarea;

        expect(() => instance.blur()).not.toThrow();
    });
});
