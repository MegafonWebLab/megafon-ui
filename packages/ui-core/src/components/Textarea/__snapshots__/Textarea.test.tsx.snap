// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Textarea /> is checked shouldComponentUpdate 1`] = `
<div
  className="mfui-textarea mfui-textarea_error mfui-textarea_color_default"
>
  <div
    className="mfui-textarea__field-wrapper"
  >
    <div>
      <textarea
        autoComplete="off"
        className="mfui-textarea__field"
      />
    </div>
  </div>
  <div
    className="mfui-textarea__text mfui-textarea__text_error"
  >
    text
  </div>
</div>
`;

exports[`<Textarea /> is checked shouldComponentUpdate with new props 1`] = `
<div
  className="mfui-textarea mfui-textarea_color_default"
>
  <div
    className="mfui-textarea__field-wrapper"
  >
    <div>
      <textarea
        autoComplete="off"
        className="mfui-textarea__field"
      />
    </div>
  </div>
</div>
`;

exports[`<Textarea /> it renders Textarea 1`] = `
<div
  className="mfui-textarea mfui-textarea_color_default"
>
  <div
    className="mfui-textarea__field-wrapper"
  >
    <div>
      <textarea
        autoComplete="off"
        className="mfui-textarea__field"
      />
    </div>
  </div>
</div>
`;

exports[`<Textarea /> it renders error icon 1`] = `
<div
  className="mfui-textarea mfui-textarea_error mfui-textarea_color_default"
>
  <div
    className="mfui-textarea__field-wrapper"
  >
    <div>
      <textarea
        autoComplete="off"
        className="mfui-textarea__field"
      />
    </div>
  </div>
  <div
    className="mfui-textarea__text mfui-textarea__text_error"
  >
    text
  </div>
</div>
`;

exports[`<Textarea /> it renders texts 1`] = `
<div
  className="mfui-textarea mfui-textarea_color_default"
>
  <div
    className="mfui-textarea__field-wrapper"
  >
    <div>
      <textarea
        autoComplete="off"
        className="mfui-textarea__field"
      />
    </div>
  </div>
  <div
    className="mfui-textarea__text mfui-textarea__text_comment"
  >
    text
  </div>
  <div
    className="mfui-textarea__text mfui-textarea__text_success"
  >
    text
  </div>
</div>
`;

exports[`<Textarea /> it renders valid icon 1`] = `
<div
  className="mfui-textarea mfui-textarea_valid mfui-textarea_color_default"
>
  <div
    className="mfui-textarea__field-wrapper"
  >
    <div>
      <textarea
        autoComplete="off"
        className="mfui-textarea__field"
      />
    </div>
  </div>
  <div
    className="mfui-textarea__text mfui-textarea__text_error"
  >
    text
  </div>
</div>
`;

exports[`<Textarea /> it renders without error icon 1`] = `
<div
  className="mfui-textarea mfui-textarea_error mfui-textarea_color_default"
>
  <div
    className="mfui-textarea__field-wrapper"
  >
    <div>
      <textarea
        autoComplete="off"
        className="mfui-textarea__field"
      />
    </div>
  </div>
</div>
`;

exports[`<Textarea /> it renders without valid icon 1`] = `
<div
  className="mfui-textarea mfui-textarea_valid mfui-textarea_color_default"
>
  <div
    className="mfui-textarea__field-wrapper"
  >
    <div>
      <textarea
        autoComplete="off"
        className="mfui-textarea__field"
      />
    </div>
  </div>
</div>
`;
